### PR TITLE
*: change @@tidb_enable_clustered_index to ON/OFF/INT_ONLY

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Problem Summary:

`@@tidb_enable_clustered_index` takes the values `on`, `off`, `int_only`, default: `int_only` 

- on: enable clustered index, only non-clustered index is supported for adding/removing primary keys, all are supported in the future
- off: off clustered index, only non-clustered index is supported for adding/removing primary keys, all supported in the future
- int_only: Behavior is consistent with 5.0-rc and below

Keep the `alter-primary-key` configuration item, which can only be used with `int_only`. 

If `CLUSTERED | NONCLUSTERED` is specified when creating a table, the optimization priority is higher than system variables and configuration items.

Mark `int_only`, `alter-primary-key` as `Deprecated` and output warning message when modifying system variable and clearly identify it in the documentation

### What is changed and how it works?

What's Changed, How it Works:

- change `tidb_enable_clustered_index` from `typeBool` to `typeEnum`
- fix tidb unit test with two replace
    - `s/.EnableClusteredIndex = true/.EnableClusteredIndex = OnClustered/`
    - `s/.EnableClusteredIndex = false/.EnableClusteredIndex = IntOnlyClustered/`
-  recovery `alter-primary-key`
-  add warn when set `int_only`
-  add test case -> [TestCreateClusteredTable](https://github.com/pingcap/tidb/pull/23529/files#diff-85f20dc5e5305b0ab0b2af4cbcc28d5dbc6c92efc5c016633d0efbbe6db6a5e0R312)
-  fix tidb-test repo test https://github.com/pingcap/tidb-test/pull/1174

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`: https://github.com/pingcap/docs-cn/pull/5834
- Need to cherry-pick to the release branch 5.0

### Check List 

Tests 

- Unit test
- Integration test https://github.com/pingcap/tidb-test/pull/1174

Side effects

- n/a

### Release note 

- No release note in this PR, but GA release-note will contain final result

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/23529)

### CODE REVIEW

**Comment by lysu** [comment]
need merge https://github.com/pingcap/tidb-test/pull/1174 first

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1523,8 +1523,14 @@ func isSingleIntPK(constr *ast.Constraint, lastCol *model.ColumnInfo) bool {
 // ShouldBuildClusteredIndex is used to determine whether the CREATE TABLE statement should build a clustered index table.
 func ShouldBuildClusteredIndex(ctx sessionctx.Context, opt *ast.IndexOption, isSingleIntPK bool) bool {
 	if opt == nil || opt.PrimaryKeyTp == model.PrimaryKeyTypeDefault {
-		return ctx.GetSessionVars().EnableClusteredIndex ||
-			(isSingleIntPK && ctx.GetSessionVars().IntPrimaryKeyDefaultAsClustered)
+		switch ctx.GetSessionVars().EnableClusteredIndex {
+		case variable.OffClustered:
```
Comment (coocood) [inline ddl/ddl_api.go:1527]: change this option to default is safer.

File: sessionctx/variable/varsutil.go [hunk 2]
```diff
@@ -339,6 +343,30 @@ func TiDBOptMultiStmt(opt string) int {
...
+
+// TiDBOptEnableClustered converts enable clustered options to ClusteredIndexDefMode.
+func TiDBOptEnableClustered(opt string) ClusteredIndexDefMode {
+	switch {
+	case strings.EqualFold(opt, IntOnly):
+		return IntOnlyClustered
+	case strings.EqualFold(opt, "ON") || opt == "1":
+		return OnClustered
+	default:
+		return OffClustered
```
Comment (coocood) [inline sessionctx/variable/varsutil.go:366]: Move IntOnly to default is better.

File: config/config.toml.example [hunk 3]
```diff
@@ -86,6 +86,12 @@ delay-clean-table-lock = 0
 # Maximum number of the splitting region, which is used by the split region statement.
 split-region-max-num = 1000
 
+# alter-primary-key is used to control alter primary key feature for the table created without special clustered/nonclustered when @@global.tidb_enable_clustered_index = 'int_only'.
+# alter-primary-key is used to control whether a primary key created without a clustered attribute under @@global.tidb_enable_clustered_index = 'int_only' can be altered.
+# Default is false, indicate single int primary will be treated as clustered and can NOT be altered, another types primary key can be altered.
+# If it is true, we can alter primary key for all types in new created table after turn it as true.
```
Comment (tangenta) [inline config/config.toml.example:92]:
```suggestion
# alter-primary-key is used to control whether the primary keys are clustered. 
# Note that this config is deprecated. Only valid when @@global.tidb_enable_clustered_index = 'int_only'.
# Default is false, only the integer primary keys are clustered.
# If it is true, all types of primary keys are nonclustered.
```

File: sessionctx/variable/varsutil.go [hunk 4]
```diff
@@ -339,6 +343,30 @@ func TiDBOptMultiStmt(opt string) int {
 	return WarnInt
 }
 
+// ClusteredIndexDefMode controls the default clustered property for primary key.
+type ClusteredIndexDefMode int
+
+const (
+	// IntOnlyClustered indicates only single int primary key will default be clustered.
+	IntOnlyClustered ClusteredIndexDefMode = 0
```
Comment (tangenta) [inline sessionctx/variable/varsutil.go:351]: I suggest to name it `ClusteredIndexDefModeIntOnly`, `ClusteredIndexDefModeOff`, `ClusteredIndexDefModeOn` for more friendly autocomplete.

Comment (lysu) [reply sessionctx/variable/varsutil.go]: renamed~

**Comment by lysu** [comment]
@tangenta PTAL

**Pull request merged**